enhancements/insights/up-to-date-gathering.md

```diff
@@ -0,0 +1,250 @@
+---
+title: insights-up-to-date-gathering
+authors:
+  - "@martinkunc"
+  - "@inecas"
+reviewers:
+  - "@smarterclayton"
+  - "@mfojtik"
+  - "@jhjaggars"
+approvers:
+  - "@smarterclayton"
+  - "@mfojtik"
+  - "@jhjaggars"
+creation-date: 2020-08-21
+last-updated: 2020-08-21
+status: provisional
+---
+
+# Up to date gathering in Insights operator
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The intent of this change is to evolve Insights Operator to make the data gathering component less dependent 
+on the the particular release of the cluster. The proposal also suggest to follow more the standard operator 
+pattern, where the responsibility of the main operator process would shift from gathering the data itself to 
+watch over the gathering process, such as the selection of desired image stream or the fallback mechanism in 
+case something goes wrong.
+
+![](io-gatherer-with-container.png)
+
+## Motivation
+
+The motivation is to be able to react faster on requirements for additional
+Insights Operator data requests. The benefit for customers would be faster
+identification of clusters based on newly discovered issues, identified by
+particular cluster configuration.
+The other potential slowdown might be caused by fact that administrators need
+to upgrade their clusters to deliver actual updates (either automatically or manually).
+
+Example of situation when the process is too slow:
+Two weeks ago we discovered an issue that for larger amount of generated alerts, IO Archives
+becomes too large (#148, CCXDEV-2673). We decided to fix it in all available versions (4.6, - 4.3)
+We spent 3 days on development and review until it was merged in Master/4.6
+3 Days later we created PR for 4.5, it was merged 6 days later.
+6 days after we creted PR for 4.4 it is still not merged (now e2e tests are failing)
+We still need to merge it to 4.3.
+Since the beginning it is now 24 days.
+
+When CCX needs release faster: 
+- To identify affected customers (suggest Workaround), 
+- To find real feature usage, especially in in historical releases (because all that waiting on merge)
+- Fix IO bug
+
+
+### Goals
+
+As part of this enhancement we plan to do the following:
+* Release process change. On a fast-track release, publish Insights Operator fast-track changes during build to new quay tag for example 4.5-fast-track.
+The rest of release process will behave like today, release to 4.x.
+
+There will be two containers in Insights Operator Pod (*main* container with driver and *gatherer* container with Insights Operator).
```

Comment:
This sounds like trying to bypass being part of the payload (which updates at a certain rate) and seems to align with a lifecycle delivered via OLM.  This proposal should probably discuss "insights-operator delivered via OLM" as an alternative to this design, which would solve those problems more effectively.

Reply:
Those are good points. Questions given we have an option for insights-operator to be delivered via OLM:

* would there be a way to have the OLM one installed by default?
* or would we keep the OLM one as optional, and still have the CVO present by default?
* or could the insights-operator run an image from OLM instead of the CVO one, but still the same pod?

Reply:
Alternatively, the insights-operator could collect payloads from external gatherers (with having it's own by default), and these gatherers being provided by OLM: it could be updated version of io itself, but there might be more, especially around other OLM operators for managed services.

Reply:
I dont know much about lifecycle of OLM, if it would allow faster updates. But combined with failback to standard gathering it makes sense.

I was thinking if we would keep current Gathering + Saving archives to volume move to a shared volume, where whoever would be interested could provide more files to upload in some schema.
For example we have insights-2020-09-15-090843.tar.gz, but any other operator could leave here "some-other-data-[DATE].tar.gz", which could main IO just upload (latest by date).

Reply:
I realised that the suggestion might look confusing when it is mixing deployment using standard CVO operated operator and also custom deployment with reaching quay manually. The reason for this was that CVO based operator would act "slow" branch where we would fail-back from "insights operator" downloaded by custom mechanism.

It looks like using CVO and OLM managed fast track would be more suitable. Would you see any problems when CVO classic operator would install operator using OLM and let it gather (eventually fail-back to CVO, released one) ?